File: src/Exporter.zig
const std = @import("std");
const builtin = @import("builtin");
const fmt = std.fmt;
const heap = std.heap;
const json = std.json;
const mem = std.mem;
const net = std.net;
const time = std.time;

const prometheus = @import("prometheus");
const sqlite = @import("sqlite");

const HealthData = @import("HealthData.zig");

const logger = std.log.scoped(.exporter);

const Self = @This();

allocator: *mem.Allocator,
db: *sqlite.Db,
addr: net.Address,

stream: ?net.Stream = null,

pub fn init(allocator: *mem.Allocator, db: *sqlite.Db, addr: net.Address) !Self {
    var res = Self{
        .allocator = allocator,
        .db = db,
        .addr = addr,
    };
    return res;
}

pub fn deinit(self: *Self) void {
    if (self.stream) |stream| {
        stream.close();
    }
}

pub fn run(self: *Self) !void {
    while (true) {
        try self.doExport();
        time.sleep(10 * time.ns_per_s);
    }
}

fn doExport(self: *Self) !void {
    const start = time.milliTimestamp();
    defer {
        logger.info("exported data in {d}ms", .{time.milliTimestamp() - start});
    }
    logger.debug("exporting data", .{});

    // Try to connect to Victoria
    self.stream = net.tcpConnectToAddress(self.addr) catch |err| {
        logger.err("unable to connect to victoria at {s}, err: {s}", .{ self.addr, err });
        return;
    };

    // Prepare the statements

    var diags = sqlite.Diagnostics{};
    var stmts = Statements.prepare(self.db, &diags) catch |err| {
        logger.err("unable to prepare statements, err: {s}, diagnostics: {s}", .{ err, diags });
        return err;
    };
    defer stmts.deinit();

    // Export

    // Assume the stream is present, otherwise we wouldn't be called;
    var stream = self.stream.?;
    var writer = stream.writer();

    try self.exportHeartRate(writer, &stmts);
    try self.exportGeneric(writer, &stmts, .weight_body_mass);
    try self.exportGeneric(writer, &stmts, .walking_heart_rate_average);
    try self.exportGeneric(writer, &stmts, .resting_heart_rate);
    try self.exportGeneric(writer, &stmts, .walking_running_distance);
    try self.exportGeneric(writer, &stmts, .walking_speed);
    try self.exportSleepAnalysis(writer, &stmts);
}

// TODO(vincent): merge both functions ? they're quite similar.

fn exportHeartRate(self: *Self, writer: net.Stream.Writer, stmts: *Statements) !void {
    // Store the data point ids to mark as exported next.
    var ids = try std.ArrayList(usize).initCapacity(self.allocator, 64);
    defer ids.deinit();

    // Export all data points
    stmts.get_heart_rate_data_point.reset();
    var iter = try stmts.get_heart_rate_data_point.iterator(
        struct {
            id: usize,
            value: f64,
            date: usize,
        },
        .{},
    );

    while (try iter.next(.{})) |row| {
        try writer.print("put health_data_heart_rate {d} {d} \n", .{
            row.date,
            row.value,
        });
        try ids.append(row.id);
    }
    logger.debug("exported {d} heart_rate data points", .{ids.items.len});

    // Mark the data point as exported

    var global_savepoint = try self.db.savepoint("global");
    defer global_savepoint.rollback();

    for (ids.items) |id| {
        stmts.mark_heart_rate_as_exported.reset();
        try stmts.mark_heart_rate_as_exported.exec(.{}, .{
            .id = id,
        });
    }

    global_savepoint.commit();
}

fn exportGeneric(self: *Self, writer: net.Stream.Writer, stmts: *Statements, comptime metric_name: @Type(.EnumLiteral)) !void {
    // Store the data point ids to mark as exported next.
    var ids = try std.ArrayList(usize).initCapacity(self.allocator, 64);
    defer ids.deinit();

    // Export all data point
    stmts.get_generic_data_point.reset();
    var iter = try stmts.get_generic_data_point.iterator(
        struct {
            id: usize,
            value: f64,
            date: usize,
        },
        .{
            .name = @as([]const u8, @tagName(metric_name)),
        },
    );

    while (try iter.next(.{})) |row| {
        try writer.print("put health_data_" ++ @tagName(metric_name) ++ " {d} {d} \n", .{
            row.date,
            row.value,
        });
        try ids.append(row.id);
    }
    logger.debug("exported {d} " ++ @tagName(metric_name) ++ " data point", .{ids.items.len});

    // Mark the data point as exported

    var global_savepoint = try self.db.savepoint("global");
    defer global_savepoint.rollback();

    for (ids.items) |id| {
        stmts.mark_generic_as_exported.reset();
        try stmts.mark_generic_as_exported.exec(.{}, .{
            .id = id,
        });
    }

    global_savepoint.commit();
}

fn exportSleepAnalysis(self: *Self, writer: net.Stream.Writer, stmts: *Statements) !void {
    // Store the data point ids to mark as exported next.
    var ids = try std.ArrayList(usize).initCapacity(self.allocator, 64);
    defer ids.deinit();

    // Export all data point
    stmts.get_sleep_analysis_data_point.reset();
    var iter = try stmts.get_sleep_analysis_data_point.iterator(
        struct {
            id: usize,
            in_bed: f64,
            asleep: f64,
            date: usize,
        },
        .{},
    );

    while (try iter.next(.{})) |row| {
        try writer.print("put health_data_sleep_analysis {d} {d} type=in_bed\n", .{
            row.date,
            row.in_bed,
        });
        try writer.print("put health_data_sleep_analysis {d} {d} type=asleep\n", .{
            row.date,
            row.asleep,
        });
        try ids.append(row.id);
    }
    logger.debug("exported {d} sleep_analysis data points", .{ids.items.len});

    // Mark the data point as exported

    var global_savepoint = try self.db.savepoint("global");
    defer global_savepoint.rollback();

    for (ids.items) |id| {
        stmts.mark_sleep_analysis_as_exported.reset();
        try stmts.mark_sleep_analysis_as_exported.exec(.{}, .{
            .id = id,
        });
    }

    global_savepoint.commit();
}

const Statements = struct {
    const get_heart_rate_data_point_query =
        \\SELECT d.id, d.max, d.date
        \\FROM data_point_heart_rate d
        \\INNER JOIN metric m ON d.metric_id = m.id
        \\WHERE m.name = 'heart_rate'
        \\AND d.exported = 0
    ;

    const get_sleep_analysis_data_point_query =
        \\SELECT d.id, d.in_bed, d.asleep, d.date
        \\FROM data_point_sleep_analysis d
        \\INNER JOIN metric m ON d.metric_id = m.id
        \\WHERE m.name = 'sleep_analysis'
        \\AND d.exported = 0
    ;

    const get_generic_data_point_query =
        \\SELECT d.id, d.quantity, d.date
        \\FROM data_point_generic d
        \\INNER JOIN metric m ON d.metric_id = m.id
        \\WHERE m.name = ?{[]const u8}
        \\AND d.exported = 0
    ;

    const mark_heart_rate_as_exported_query =
        \\UPDATE data_point_heart_rate
        \\SET exported = 1
        \\WHERE id = ?{usize}
    ;
    const mark_sleep_analysis_as_exported_query =
        \\UPDATE data_point_sleep_analysis
        \\SET exported = 1
        \\WHERE id = ?{usize}
    ;
    const mark_generic_as_exported_query =
        \\UPDATE data_point_generic
        \\SET exported = 1
        \\WHERE id = ?{usize}
    ;

    get_heart_rate_data_point: sqlite.StatementType(.{}, get_heart_rate_data_point_query),
    get_sleep_analysis_data_point: sqlite.StatementType(.{}, get_sleep_analysis_data_point_query),
    get_generic_data_point: sqlite.StatementType(.{}, get_generic_data_point_query),
    mark_heart_rate_as_exported: sqlite.StatementType(.{}, mark_heart_rate_as_exported_query),
    mark_sleep_analysis_as_exported: sqlite.StatementType(.{}, mark_sleep_analysis_as_exported_query),
    mark_generic_as_exported: sqlite.StatementType(.{}, mark_generic_as_exported_query),

    fn prepare(db: *sqlite.Db, diags: *sqlite.Diagnostics) !Statements {
        var res: Statements = undefined;

        res.get_heart_rate_data_point = try db.prepareWithDiags(get_heart_rate_data_point_query, .{
            .diags = diags,
        });
        res.get_sleep_analysis_data_point = try db.prepareWithDiags(get_sleep_analysis_data_point_query, .{
            .diags = diags,
        });
        res.get_generic_data_point = try db.prepareWithDiags(get_generic_data_point_query, .{
            .diags = diags,
        });
        res.mark_heart_rate_as_exported = try db.prepareWithDiags(mark_heart_rate_as_exported_query, .{
            .diags = diags,
        });
        res.mark_sleep_analysis_as_exported = try db.prepareWithDiags(mark_sleep_analysis_as_exported_query, .{
            .diags = diags,
        });
        res.mark_generic_as_exported = try db.prepareWithDiags(mark_generic_as_exported_query, .{
            .diags = diags,
        });

        return res;
    }

    fn deinit(self: *Statements) void {
        self.get_heart_rate_data_point.deinit();
        self.get_generic_data_point.deinit();
        self.mark_heart_rate_as_exported.deinit();
        self.mark_generic_as_exported.deinit();
    }
};

File: src/exporter.rs
use crate::db;
use crate::shutdown::Shutdown;
use std::fmt;
use std::fmt::Write;
use std::io;
use std::net;
use std::sync::Arc;
use std::time;
use tokio::io::AsyncWriteExt;
use tokio::net::TcpStream;
use tracing::{debug, error, info};

#[derive(Debug, thiserror::Error)]
pub enum Error {
    #[error(transparent)]
    IO(#[from] io::Error),
    #[error(transparent)]
    SQLx(#[from] sqlx::Error),
    #[error(transparent)]
    Fmt(#[from] fmt::Error),
}

pub type Result<T> = std::result::Result<T, Error>;

pub struct Exporter {
    db: Arc<db::Db>,
    addr: net::SocketAddr,
    stream: Option<TcpStream>,
}

impl Exporter {
    pub fn new(db: Arc<db::Db>, addr: net::SocketAddr) -> Self {
        Self {
            db,
            addr,
            stream: None,
        }
    }

    pub async fn run(mut self, mut shutdown: Shutdown) -> Result<()> {
        let mut interval = tokio::time::interval(time::Duration::from_secs(1));

        'outer_loop: loop {
            tokio::select! {
                _ = shutdown.recv() => {
                    info!("exporter shutting down");
                    break 'outer_loop;
                },
                _ = interval.tick() => {
                    match self.do_export().await {
                        Ok(_) => {},
                        Err(err) => error!(%err, "unable to export data"),
                    }
                },
            }
        }

        if let Some(stream) = self.stream {
            drop(stream);
        }

        Ok(())
    }

    async fn connect(&mut self) -> Result<&mut TcpStream> {
        match self.stream {
            Some(ref mut s) => Ok(s),
            None => {
                debug!(addr = self.addr.to_string(), "connecting to VM");

                let stream = TcpStream::connect(self.addr).await?;
                self.stream = Some(stream);

                // Safe because we know it's there
                Ok(self.stream.as_mut().unwrap())
            }
        }
    }

    async fn do_export(&mut self) -> Result<()> {
        let mut exported: usize = 0;
        let mut commands_buffer = String::new();

        // Generate all the export commands

        self.gen_export_heart_rate_commands(&mut commands_buffer, &mut exported)
            .await?;
        for metric_type in ALL_GENERIC_METRIC_TYPES {
            self.gen_export_generic_commands(&mut commands_buffer, metric_type, &mut exported)
                .await?;
        }
        self.gen_export_sleep_analysis_commands(&mut commands_buffer, &mut exported)
            .await?;

        // Send them to Victoria
        let stream = self.connect().await?;
        stream.write_all(commands_buffer.as_bytes()).await?;

        if exported > 0 {
            info!(exported = exported, "exported data points");
        }

        Ok(())
    }

    async fn gen_export_heart_rate_commands(
        &mut self,
        commands_buffer: &mut String,
        exported: &mut usize,
    ) -> Result<()> {
        // Get the data and write the export commands to the buffer

        let rows = sqlx::query!(
            r#"
            SELECT d.id, d.max, d.date
            FROM data_point_heart_rate d
            INNER JOIN metric m ON d.metric_id = m.id
            WHERE m.name = 'heart_rate'
            AND d.exported = false"#,
        )
        .fetch_all(&self.db.pool)
        .await?;

        let mut ids = Vec::<i64>::new();
        for row in rows {
            writeln!(
                commands_buffer,
                "put health_data_heart_rate {} {} ",
                row.date.unix_timestamp_nanos() / 1_000_000,
                row.max
            )?;

            ids.push(row.id);
        }

        // Mark all data points as exported

        *exported += ids.len();

        let mut tx = self.db.pool.begin().await?;
        for id in ids {
            sqlx::query!(
                r#"
                UPDATE data_point_heart_rate
                SET exported = true
                WHERE id = $1"#,
                id,
            )
            .execute(&mut tx)
            .await?;
        }
        tx.commit().await?;

        Ok(())
    }

    async fn gen_export_generic_commands(
        &mut self,
        commands_buffer: &mut String,
        metric_type: GenericMetricType,
        exported: &mut usize,
    ) -> Result<()> {
        let metric_name = metric_type.to_string();

        // Get the data and write the export commands to the buffer

        let rows = sqlx::query!(
            r#"
            SELECT d.id, d.quantity, d.date
            FROM data_point_generic d
            INNER JOIN metric m ON d.metric_id = m.id
            WHERE m.name = $1
            AND d.exported = false
            "#,
            metric_name,
        )
        .fetch_all(&self.db.pool)
        .await?;

        let mut ids = Vec::<i64>::new();
        for row in rows {
            writeln!(
                commands_buffer,
                "put health_data_{} {} {} ",
                metric_name,
                row.date.unix_timestamp_nanos() / 1_000_000,
                row.quantity
            )?;

            ids.push(row.id);
        }

        // Mark all data points as exported

        *exported += ids.len();

        let mut tx = self.db.pool.begin().await?;
        for id in ids {
            sqlx::query!(
                r#"
                UPDATE data_point_generic
                SET exported = true
                WHERE id = $1"#,
                id,
            )
            .execute(&mut tx)
            .await?;
        }
        tx.commit().await?;

        Ok(())
    }

    async fn gen_export_sleep_analysis_commands(
        &mut self,
        commands_buffer: &mut String,
        exported: &mut usize,
    ) -> Result<()> {
        // Get the data and write the export commands to the buffer

        let rows = sqlx::query!(
            r#"
            SELECT d.id, d.in_bed, d.asleep, d.date
            FROM data_point_sleep_analysis d
            INNER JOIN metric m ON d.metric_id = m.id
            WHERE m.name = 'sleep_analysis'
            AND d.exported = false"#,
        )
        .fetch_all(&self.db.pool)
        .await?;

        let mut ids = Vec::<i64>::new();
        for row in rows {
            writeln!(
                commands_buffer,
                "put health_data_sleep_analysis {} {} type=in_bed",
                row.date.unix_timestamp_nanos() / 1_000_000,
                row.in_bed,
            )?;
            writeln!(
                commands_buffer,
                "put health_data_sleep_analysis {} {} type=asleep",
                row.date.unix_timestamp_nanos() / 1_000_000,
                row.asleep,
            )?;

            ids.push(row.id);
        }

        // Mark all data points as exported

        *exported += ids.len();

        let mut tx = self.db.pool.begin().await?;
        for id in ids {
            sqlx::query!(
                r#"
                UPDATE data_point_sleep_analysis
                SET exported = true
                WHERE id = $1"#,
                id,
            )
            .execute(&mut tx)
            .await?;
        }
        tx.commit().await?;

        Ok(())
    }
}

const ALL_GENERIC_METRIC_TYPES: [GenericMetricType; 5] = [
    GenericMetricType::WeightBodyMass,
    GenericMetricType::WalkingHeartRateAverage,
    GenericMetricType::RestingHeartRate,
    GenericMetricType::WalkingRunningDistance,
    GenericMetricType::WalkingSpeed,
];

enum GenericMetricType {
    WeightBodyMass,
    WalkingHeartRateAverage,
    RestingHeartRate,
    WalkingRunningDistance,
    WalkingSpeed,
}

impl fmt::Display for GenericMetricType {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Self::WeightBodyMass => write!(f, "weight_body_mass"),
            Self::WalkingHeartRateAverage => write!(f, "walking_heart_rate_average"),
            Self::RestingHeartRate => write!(f, "resting_heart_rate"),
            Self::WalkingRunningDistance => write!(f, "walking_running_distance"),
            Self::WalkingSpeed => write!(f, "walking_speed"),
        }
    }
}
